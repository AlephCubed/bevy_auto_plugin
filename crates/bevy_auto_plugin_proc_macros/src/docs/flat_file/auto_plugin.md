The main attribute for flat file mode that processes all auto attributes in the file.

# Parameters
- `app=identifier` - Optional. Specifies the app variable name that code will be injected into. (defaults to app)

# Example
```rust
use bevy::prelude::*;
use bevy_auto_plugin::modes::flat_file::prelude::*;

// Example attributes or declarations for components, events, or resources
// #[auto_register_type]
// #[derive(Component, Reflect)]
// #[reflect(Component)]
// struct MyComponent;

// ^ auto macro attributes must be declared above #[auto_plugin]
#[auto_plugin(app=app)]
fn plugin(app: &mut App) {
    // Code generated by the macro is injected here.
    // For example:
    // app.register_type::<MyComponent>();

    // Your custom logic comes here.
}
```