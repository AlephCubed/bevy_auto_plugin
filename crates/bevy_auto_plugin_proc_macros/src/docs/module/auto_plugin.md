The main attribute for module mode that processes all auto attributes in the module.

# Parameters
- `init_name=identifier` - Optional. Specifies the name of the generated function that initializes the plugin. (defaults to init)

# Example
```rust
use bevy::prelude::*;
use bevy_auto_plugin::module::prelude::*;

#[auto_plugin(init_name=init)]
pub mod my_plugin {
    use bevy::prelude::*;
    use bevy_auto_plugin::module::prelude::*;

    #[auto_register_type]
    #[derive(Component, Reflect)]
    #[reflect(Component)]
    pub struct MyComponent;

    /* code gen */
    // pub(super) fn init(app: &mut App) {  
    //     app.register_type::<MyComponent>();
    // }
}

fn plugin(app: &mut App) {
    app.add_plugins(my_plugin::init);
}
```