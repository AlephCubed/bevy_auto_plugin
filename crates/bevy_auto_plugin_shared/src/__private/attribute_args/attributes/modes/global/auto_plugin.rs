use crate::__private::attribute_args::GenericsArgs;
use crate::__private::type_list::TypeList;
use darling::FromMeta;
use proc_macro2::Ident;
use syn::Path;

#[derive(FromMeta, Debug, Default, Clone, PartialEq)]
#[darling(derive_syn_parse, default)]
pub struct AutoPluginStructOrEnumAttributeArgs {
    #[darling(multiple)]
    pub generics: Vec<TypeList>,
    pub impl_plugin_trait: bool,
    pub impl_generic_auto_plugin_trait: bool,
    pub impl_generic_plugin_trait: bool,
}

impl GenericsArgs for AutoPluginStructOrEnumAttributeArgs {
    fn type_lists(&self) -> &[TypeList] {
        &self.generics
    }
}

#[derive(FromMeta, Debug, Default, Clone, PartialEq)]
#[darling(derive_syn_parse, default)]
pub struct AutoPluginFnAttributeArgs {
    #[darling(multiple)]
    pub generics: Vec<TypeList>,
    pub plugin: Option<Path>,
    pub app_param: Option<Ident>,
}

impl GenericsArgs for AutoPluginFnAttributeArgs {
    const TURBOFISH: bool = true;
    fn type_lists(&self) -> &[TypeList] {
        &self.generics
    }
}
