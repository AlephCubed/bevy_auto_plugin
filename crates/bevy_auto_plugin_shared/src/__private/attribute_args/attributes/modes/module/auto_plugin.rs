use crate::__private::attribute_args::GenericsArgs;
use crate::__private::type_list::TypeList;
use darling::FromMeta;
use proc_macro2::Ident;
use proc_macro2::Span;
use smart_default::SmartDefault;

#[derive(FromMeta, Debug, SmartDefault, Clone, PartialEq)]
#[darling(derive_syn_parse, default)]
pub struct AutoPluginArgs {
    #[darling(multiple)]
    pub generics: Vec<TypeList>,
    #[default(Ident::new("init", Span::call_site()))]
    pub init_name: Ident,
}

impl GenericsArgs for AutoPluginArgs {
    const TURBOFISH: bool = true;
    fn type_lists(&self) -> &[TypeList] {
        &self.generics
    }
}
