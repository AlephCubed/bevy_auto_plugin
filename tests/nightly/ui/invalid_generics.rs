use bevy_auto_plugin_macros::auto_plugin::*;


#[auto_register_type(Test)]
struct Test<T>(T);

#[auto_plugin(app=_app)]
fn plugin(_app: &mut bevy_app::App) {}

// dummy main
fn main() {
    
}