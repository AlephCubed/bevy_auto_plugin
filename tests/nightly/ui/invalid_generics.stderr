error: expected 1 generic parameters, found 0
 --> tests/nightly/ui/invalid_generics.rs:4:22
  |
4 | #[auto_register_type(Test)]
  |                      ^^^^
